src/functional.rs

```diff
@@ -279,6 +283,77 @@ impl<K, V> Node<K, V> {
             }
         }
     }
+
+    /// Returns a new tree by rotating the right child up to be come the root. To maintain the BST
+    /// invariant we lift the right child's left child to be the new left child's right child.
+    /// root.
+    fn rotate_left(&self) -> Self {
+        match self.right.as_ref() {
+            Tree::Leaf => self.clone(),
+            Tree::Node(r) => {
+                let new_left = Tree::Node(Self {
+                    height: self.left.height().max(r.left.height()) + 1,
+                    key: Rc::clone(&self.key),
+                    left: Rc::clone(&self.left),
+                    right: Rc::clone(&r.left),
+                    value: Rc::clone(&self.value),
+                });
+
+                Self {
+                    height: r.right.height().max(new_left.height()) + 1, // is max necessary here?
+                    key: Rc::clone(&r.key),
+                    left: Rc::new(new_left),
+                    right: Rc::clone(&r.right),
+                    value: Rc::clone(&r.value),
+                }
+            }
+        }
+    }
+
+    /// Returns a new tree by rotating the left child up to be come the root. To maintain the BST
+    /// invariant we lift the left child's right child to be the new right child's left child.
+    fn rotate_right(&self) -> Self {
+        match self.left.as_ref() {
+            Tree::Leaf => self.clone(),
+            Tree::Node(l) => {
+                let new_right = Tree::Node(Self {
+                    height: self.right.height().max(l.right.height()) + 1,
+                    key: Rc::clone(&self.key),
+                    left: Rc::clone(&l.right),
+                    right: Rc::clone(&self.right),
+                    value: Rc::clone(&self.value),
+                });
+
+                Self {
+                    height: l.left.height().max(new_right.height()) + 1,
+                    key: Rc::clone(&l.key),
+                    left: Rc::clone(&l.left),
+                    right: Rc::new(new_right),
+                    value: Rc::clone(&l.value),
+                }
+            }
+        }
+    }
+
+    /// Balances a tree using the heights of the children.
+    ///
+    /// **Note** This takes `self` instead of `&self` which isn't very functional one might say
+    /// but it isn't `mut` and it saves us from an unnecessary `clone` when it's already balanced.
+    fn balance(self) -> Self {
+        let return_node = if self.right.height() > self.left.height() + 1 {
+            self.rotate_left()
+        } else if self.left.height() > self.right.height() + 1 {
+            self.rotate_right()
+        } else {
+            self
+        };
+        if cfg!(test) {
```

Comment:
At compile time for non-tests this should become `if false` and get compiled out. In tests it's `if true` and stays in. The reason I chose this instead of `#[cfg(test)]` is because that removes the branch before doing linting and causes Rust to say, "You're assigning `return_node` and then immediately returning it - just return it directly!". This `if cfg!(test)` means that the ranch survives a few more steps but is hopefully removed by someone down the line.